### PR TITLE
Distributed s3 delete objects

### PR DESCRIPTION
**Pull request by malachi-constant**
### Feature or Bugfix
- Refactor `s3.delete_objects` to run in distributed fashion.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: da502a365c3011e035c1b82303f9d20edef55603
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ea551a9c-28e9-4333-a7d6-d6e81e19a004%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 885a15c39e9417e353c09eaaeace03da4e899e1a
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0b8f6a86-a695-431f-955e-0dab0487c028%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: fd9ad263ada2b5346512eceba6d5d461f23bbd8e
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=44d06bc8-6a3c-4f3f-a01e-9fe87265b9af%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: d71b6718b9466cb86dad617bfddee4ba0942e1ea
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=59f9d320-1a1d-4532-910c-d3c211650d01%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: tests/test_s3.py [hunk 1]
```diff
@@ -69,6 +69,17 @@ def test_list_by_last_modified_date(path):
     assert len(wr.s3.read_json(path, last_modified_begin=begin_utc, last_modified_end=end_utc).index) == 6
 
 
+def test_s3_delete_objects(path):
```
Comment (jaidisido) [inline tests/test_s3.py:72]: Perhaps we can move this to load tests and try with a much higher number of files?

File: awswrangler/s3/_delete.py [hunk 2]
```diff
@@ -146,29 +118,11 @@ def delete_objects(
...
-                list(
-                    executor.map(
-                        _delete_objects_concurrent,
-                        itertools.repeat(bucket),
-                        chunks,
-                        itertools.repeat(s3_additional_kwargs),
-                        itertools.repeat(_utils.boto3_to_primitives(boto3_session=boto3_session)),
-                    )
-                )
+        executor = _get_executor(use_threads=use_threads)
```
Comment (jaidisido) [inline awswrangler/s3/_delete.py:125]: So a `ray_get` was not required in the end? Not sure how this operation would be blocking without it?

**Review by jaidisido — CHANGES_REQUESTED** [review]

File: awswrangler/s3/_delete.py [hunk 3]
```diff
@@ -49,36 +51,6 @@ def _delete_objects(
     deleted: List[Dict[str, Any]] = res.get("Deleted", [])
     for obj in deleted:
         _logger.debug("s3://%s/%s has been deleted.", bucket, obj.get("Key"))
-    errors: List[Dict[str, Any]] = res.get("Errors", [])
```
[the remark this answers is not in the document]
Comment (malachi-constant) [reply awswrangler/s3/_delete.py:52]: Agreed, let me figure out a way to mock this..

File: awswrangler/s3/_delete.py [hunk 4]
```diff
@@ -146,29 +118,11 @@ def delete_objects(
         last_modified_end=last_modified_end,
         s3_additional_kwargs=s3_additional_kwargs,
     )
-    if len(paths) < 1:
-        return
+
     buckets: Dict[str, List[str]] = _split_paths_by_bucket(paths=paths)
     for bucket, keys in buckets.items():
```
[the remark this answers is not in the document]
Comment (malachi-constant) [reply awswrangler/s3/_delete.py:123]: Agreed.

[the remark this answers is not in the document]
Comment (malachi-constant) [reply awswrangler/s3/_delete.py:123]: Hmm I'm wondering what the best way to do that is, the underlying call [s3.delete_objects()](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html?highlight=delete#S3.Client.delete_objects) expects the batch of keys to map to a _single_ bucket.

[the remark this answers is not in the document]
Comment (jaidisido) [reply awswrangler/s3/_delete.py:123]: True, I guess the way to do it is to refactor how chunks are build to also include the bucket name. At which point, the bucket vs path can be resolved inside the remote function and you wouldn't need to pass bucket to the executor anymore

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 87a92e2bbdc92d5a24528cfb982bbe00e3f96b39
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8d6eda49-f433-4236-bdd6-0cdcf0091058%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/s3/_delete.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (malachi-constant) [reply awswrangler/s3/_delete.py]: Ok, I think it should be much cleaner now.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: babf83553bec5a0bd963285d24e2c80d93249a0c
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a5b2852c-e76f-4658-a5ca-20db1bbe72fd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: abb22e678d65373a459fd53bcf82362aa65c950f
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a1cf05c1-f40e-4877-952f-3645c0de3ccb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 13308c5808fb97b1b1bedfcade0f23ec67fb0092
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0c6a7491-3a1f-4997-b192-e1948c3ba103%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 9fd3f4c2f9b2cd52b7b4d96da1bbc4e06c4209dc
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=722d1c13-1ff6-4c6f-b4b1-dedab22c827a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 9b85b44f8d87decf26fad399327ae36c3093b6c0
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1efebbff-dd0e-4244-a745-ee7f41e33246%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/s3/_delete.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (malachi-constant) [reply awswrangler/s3/_delete.py:52]: So this is only going to handle retries for `5xx` errors. There are many possible [error codes](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html?highlight=delete#S3.Client.delete_objects), and most `4xx` would not make sense to retry but I'm worried I'm not accounting for some edge case. 

Thoughts @jaidisido ?

File: awswrangler/s3/_delete.py [hunk 5]
```diff
@@ -1,16 +1,15 @@
-"""Amazon S3 CopDeletey Module (PRIVATE)."""
+"""Amazon S3 Delete Module (PRIVATE)."""
```
Comment (jaidisido) [inline awswrangler/s3/_delete.py:1]: It looks so much cleaner, well done!

File: awswrangler/s3/_delete.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jaidisido) [reply awswrangler/s3/_delete.py:52]: I think this is a better approach, it wouldn't make sense to retry on a 403 for example as you say

File: awswrangler/s3/_delete.py [hunk 6]
```diff
@@ -146,29 +117,18 @@ def delete_objects(
...
-                        chunks,
-                        itertools.repeat(s3_additional_kwargs),
-                        itertools.repeat(_utils.boto3_to_primitives(boto3_session=boto3_session)),
-                    )
-                )
+    paths_by_bucket: Dict[str, List[str]] = _split_paths_by_bucket(paths)
+
+    chunks = []
+    for _, paths in paths_by_bucket.items():
+        chunks += _utils.chunkify(lst=paths, max_length=100)
```
Comment (jaidisido) [inline awswrangler/s3/_delete.py:124]: Any reason why `max_length` was reduced to 100 instead of 1_000?

File: load_tests/_utils.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import random
```
Comment (jaidisido) [inline load_tests/_utils.py:1]: Looks like we will be duplicating a lot of the code here with standard tests. I think we should consider merging the two directories. But let us do that in a separate PR

File: load_tests/test_s3.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    end = time.time()
+
+    elapsed_time = end - start
+    assert elapsed_time < benchmark_time
+
+
+def test_s3_delete_objects(path, path2):
+    df = pd.DataFrame({"id": [1, 2, 3]})
+    objects_per_bucket = 505
+    paths = [f"s3://{path}delete-test{i}.json" for i in range(objects_per_bucket)] + [
```
Comment (jaidisido) [inline load_tests/test_s3.py:31]: doesn't `path` already contain `s3://`?

File: load_tests/test_s3.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        f"s3://{path2}delete-test{i}.json" for i in range(objects_per_bucket)
+    ]
+    for path in paths:
+        wr.s3.to_json(df, path)
+    start = time.time()
+    wr.s3.delete_objects(path=paths)
+    end = time.time()
+    print(f"elapsed time: {end - start}")
+    assert len(wr.s3.list_objects(f"{path}delete-test*")) == 0
```
Comment (jaidisido) [inline load_tests/test_s3.py:40]: To be on the safe side here perhaps we should just store the lists of paths in two separate variables and pass them here:
```
paths1 = [f"s3://{path}delete-test{i}.json" for i in range(objects_per_bucket)]
paths2 = [f"s3://{path2}delete-test{i}.json" for i in range(objects_per_bucket)]
paths = paths1 + paths2
...
assert len(wr.s3.list_objects(paths1)) == 0
```

File: load_tests/test_s3.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    objects_per_bucket = 505
+    paths = [f"s3://{path}delete-test{i}.json" for i in range(objects_per_bucket)] + [
+        f"s3://{path2}delete-test{i}.json" for i in range(objects_per_bucket)
+    ]
+    for path in paths:
+        wr.s3.to_json(df, path)
+    start = time.time()
+    wr.s3.delete_objects(path=paths)
+    end = time.time()
+    print(f"elapsed time: {end - start}")
```
Comment (jaidisido) [inline load_tests/test_s3.py:39]: Did we make a decision on using `--durations=0` instead of repeating `time.time()` everywhere?

https://howchoo.com/python/how-to-measure-unit-test-execution-times-in-pytest

File: load_tests/_utils.py (hunk 7 above)
Comment (malachi-constant) [reply load_tests/_utils.py:1]: #1498

File: load_tests/test_s3.py (hunk 9 above)
Comment (malachi-constant) [reply load_tests/test_s3.py:40]: You're right, looks like `_path2list()` was allowing me to be sloppy.

File: load_tests/test_s3.py (hunk 10 above)
Comment (malachi-constant) [reply load_tests/test_s3.py:39]: You're right, looks like _path2list() was allowing me to be sloppy.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 0affa4f7f70128f95afb1949ae35b0090551b939
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5f4ec6c6-6f64-4d44-8580-981f8f7430a3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: load_tests/test_s3.py (hunk 10 above)
Comment (malachi-constant) [reply load_tests/test_s3.py:39]: This is an interesting case, because the file writes in the beginning of the test unfortunately take a lot longer than the deletes themselves. So we need some way to time just the delete.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 9548826a907c9b4c12581c378e882f681d68473a
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d2d9e81d-9044-4822-8653-430536229b2e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6d861431-7805-48e8-a7ad-9821b9c1d6c3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/s3/_delete.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (malachi-constant) [reply awswrangler/s3/_delete.py:52]: Agreed, so I've struck out on a reliable way to test this. The [docs](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#checking-retry-attempts-in-your-client-logs) seem to only suggest reading the logs at the botocore level. I also went down a rabbit hole with the [botocore Stubber](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/stubber.html) but that I don't think can just give us one "fake" request and then let the client hit the service again.

We will want to consider setting `standard` mode as a [default for wrangler](https://github.com/awslabs/aws-data-wrangler/blob/main/awswrangler/_utils.py#L66) botocore config. But i'll leave that out of this PR for now.

**Comment by malachi-constant** [comment]
Also, I need to figure out what's going on in our Load Tests job.

1) It's not failing the Codebuild run on pytest failures.
2) The delete test is failing due to setup with fixtures required by the test.
```
[31m[1m___________________ ERROR at setup of test_s3_delete_objects ___________________[0m

    @pytest.fixture(scope="session")
    def cloudformation_outputs():
>       return extract_cloudformation_outputs()

[1m[31mload_tests/conftest.py[0m:8: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[1m[31mload_tests/_utils.py[0m:16: in extract_cloudformation_outputs
    client = boto3.client("cloudformation")
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/boto3/__init__.py[0m:92: in client
    return _get_default_session().client(*args, **kwargs)
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/boto3/session.py[0m:299: in client
    return self._session.create_client(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/session.py[0m:950: in create_client
    client = client_creator.create_client(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/client.py[0m:123: in create_client
    client_args = self._get_client_args(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/client.py[0m:466: in _get_client_args
    return args_creator.get_client_args(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/args.py[0m:87: in get_client_args
    final_args = self.compute_client_args(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/args.py[0m:183: in compute_client_args
    endpoint_config = self._compute_endpoint_config(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/args.py[0m:278: in _compute_endpoint_config
    return self._resolve_endpoint(**resolve_endpoint_kwargs)
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/args.py[0m:381: in _resolve_endpoint
    return endpoint_bridge.resolve(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/client.py[0m:566: in resolve
    resolved = self.endpoint_resolver.construct_endpoint(
[1m[31manaconda3/envs/ray_cp38/lib/python3.8/site-packages/botocore/regions.py[0m:205: in construct_endpoint
    result = self._endpoint_for_partition(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.regions.EndpointResolver object at 0x7f24687f53a0>
partition = OrderedDict([('defaults', OrderedDict([('hostname', '{service}.{region}.{dnsSuffix}'), ('protocols', ['https']), ('sig...ict([('variants', [OrderedDict([('hostname', 'xray-fips.us-west-2.amazonaws.com'), ('tags', ['fips'])])])]))]))]))]))])
service_name = 'cloudformation', region_name = None
use_dualstack_endpoint = None, use_fips_endpoint = None, force_partition = False

    def _endpoint_for_partition(
        self,
        partition,
        service_name,
        region_name,
        use_dualstack_endpoint,
        use_fips_endpoint,
        force_partition=False,
    ):
        partition_name = partition["partition"]
        if (
            use_dualstack_endpoint
            and partition_name in self._UNSUPPORTED_DUALSTACK_PARTITIONS
        ):
            error_msg = (
                "Dualstack endpoints are currently not supported"
                " for %s partition" % partition_name
            )
            raise EndpointVariantError(tags=['dualstack'], error_msg=error_msg)
    
        # Get the service from the partition, or an empty template.
        service_data = partition['services'].get(
            service_name, DEFAULT_SERVICE_DATA
        )
        # Use the partition endpoint if no region is supplied.
        if region_name is None:
            if 'partitionEndpoint' in service_data:
                region_name = service_data['partitionEndpoint']
            else:
>               raise NoRegionError()
[1m[31mE               botocore.exceptions.NoRegionError: You must specify a region.[0m
```

Looks like I need to pass `AWS_DEFAULT_REGION` to ray or something similar.

File: load_tests/test_s3.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    paths2 = [f"{path2}delete-test{i}.json" for i in range(objects_per_bucket)]
+    paths = paths1 + paths2
+    for path in paths:
+        wr.s3.to_json(df, path)
+    start = time.time()
+    wr.s3.delete_objects(path=paths)
+    end = time.time()
+    print(f"elapsed time: {end - start}")
+    assert len(wr.s3.list_objects(f"{path}delete-test*")) == 0
+    assert len(wr.s3.list_objects(f"{path2}delete-test*")) == 0
```
Comment (jaidisido) [inline load_tests/test_s3.py:41]: So the reason I mention splitting paths into paths1 and paths2 is so we can use them here:
```
    assert len(wr.s3.list_objects(paths1)) == 0
    assert len(wr.s3.list_objects(paths2)) == 0
```

Comment (malachi-constant) [reply load_tests/test_s3.py:41]: As list_objects() does not support a list of paths we will leave this as is for now.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ee9a4233-bcd1-4678-b5b4-6859236184e3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0ecd7d17-49c1-440d-95c2-48ee7f534095%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3175cdca-14ee-4fde-ac51-e781e5493abf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3736d4d3-32d1-48f8-bb93-92cda4045142%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: FAILED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=388cea71-240a-4bd9-b920-4e76c09987e6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: FAILED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a956184b-c030-493d-ada6-afbc99392fef%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: b1995de452b0a52e0612c21b555300d0d0204951
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=21596a3f-eb9a-42ac-8df2-1d4f06a6226a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: load_tests/test_s3.py (hunk 10 above)
Comment (LeonLuttenberger) [reply load_tests/test_s3.py:39]: Another neat way of writing timers in Python I've used in the past is to extract the timer functionality into a context manager, which can then be used in a `with` block.

```python
from timeit import default_timer as timer

class ExecutionTimer():
    def __init__(self, msg):
        self.msg = msg
    def __enter__(self):
        self.before = timer()
    def __exit__(self, type, value, traceback):
        print(f"{self.msg}: {(timer() - self.before):.3f} sec")

with ExecutionTimer("elapsed time for wr.s3.delete_objects"):
    wr.s3.delete_objects(path=paths)
```

This block of code will then print the following:
```
elapsed time for wr.s3.delete_objects: 12.342 sec
```

Comment (malachi-constant) [reply load_tests/test_s3.py:39]: That's a nice idea, that way we could pull out the value out to do the assertion easily.

Comment (jaidisido) [reply load_tests/test_s3.py:39]: I like it, let's give it a shot plz

Comment (malachi-constant) [reply load_tests/test_s3.py]: Implemented, thanks for the tip @LeonLuttenberger , will just need to confirm benchmarks are correct now.

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: ccdaf0e36ab289271c41a6f80038130b8e8d57cf
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d222fd29-75a8-448e-a8f0-1a1175baa04c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by malachi-constant** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubLoadTests5656BB24-s6u9F3qN9oFy
* Commit ID: 71a6534e9cb0f21657d1bbb483826308f885e301
* Result: SUCCEEDED
* [Build Logs](https://pciyom2pbh.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2ff88acc-c64e-4c63-96af-81ee1f4af7c6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by jaidisido — APPROVED** [review]
